ratelimit/redis.go

```diff
@@ -0,0 +1,82 @@
+package ratelimit
+
+import (
+	"context"
+	_ "embed"
+	"strconv"
+	"time"
+
+	"github.com/redis/rueidis"
+)
+
+//go:embed gcra_rate_limit.lua
+var gcraRateLimitScript string
+
+// RedisRateLimiter is a type that represents a rate limiter that uses Redis as its backend.
+// The rate limiting is a leaky bucket implementation using the generic cell rate algorithm.
+// See https://en.wikipedia.org/wiki/Generic_cell_rate_algorithm for details on how this algorithm works.
+type RedisRateLimiter struct {
+	client rueidis.Client
+}
+
+// Ensure RedisRateLimiter implements the SharedRateLimiter interface.
+var _ SharedRateLimiter = (*RedisRateLimiter)(nil)
```

Comment:
`main` assigns it because of implementation details that could change at any point. Here this verification is part of the design and should never be removed or changed.